concordium-consensus/src/Concordium/GlobalState/Persistent/Accounts.hs

```diff
@@ -69,96 +123,123 @@ import Concordium.Types.HashableTo
+      -- | An in-memory difference map used keeping track of accounts
+      --  added in live blocks.
+      --  This is @Nothing@ if either the block is persisted or no accounts have been
+      --  added in the block (and it is thawed).
+      --  Otherwise if the block is not persisted and accounts have been added, then
+      --  the 'DiffMap.DifferenceMap' yields the @AccountAddress -> AccountIndex@ mapping for
+      --  accounts created in the block.
+      accountDiffMap :: !(IORef (Maybe DiffMap.DifferenceMap))
     }
 
+instance (IsProtocolVersion pv) => Show (Accounts pv) where
+    show accts = "Accounts: " <> show (accountTable accts)
+
 -- | A constraint that ensures a monad @m@ supports the persistent account operations.
 --  This essentially requires that the monad support 'MonadBlobStore', and 'MonadCache' for
---  the account cache.
+--  the account cache and 'MonadAccountMapStore' for the persistent account map.
 type SupportsPersistentAccount pv m =
     ( IsProtocolVersion pv,
       MonadBlobStore m,
-      MonadCache (AccountCache (AccountVersionFor pv)) m
+      MonadCache (AccountCache (AccountVersionFor pv)) m,
+      LMDBAccountMap.MonadAccountMapStore m
     )
 
-instance (IsProtocolVersion pv) => Show (Accounts pv) where
-    show a = show (accountTable a)
-
 instance (SupportsPersistentAccount pv m) => MHashableTo m H.Hash (Accounts pv) where
     getHashM Accounts{..} = getHashM accountTable
 
+-- Note. We're writing to the LMDB accountmap as part of the 'storeUpdate' implementation below.
+-- This in turn means that no associated metadata is being written to the LMDB database (i.e. the block hash) of the
+-- persisted block. If we need this, then the write to the LMDB database could be done in 'saveBlockState' and the 'DifferenceMap' should be retained up
+-- to that point.
+-- It shouldn't be necessary with this additional metadata as when (potentially) certified blocks are being rolled back, so are the
+-- accounts created in those, in turn this means that the LMDB account map will have entries for all accounts present in the last finalized block.
 instance (SupportsPersistentAccount pv m) => BlobStorable m (Accounts pv) where
     storeUpdate Accounts{..} = do
-        (pMap, accountMap') <- storeUpdate accountMap
         (pTable, accountTable') <- storeUpdate accountTable
         (pRegIdHistory, regIdHistory') <- storeUpdate accountRegIdHistory
+        mAccDiffMap <- liftIO $ readIORef accountDiffMap
+        case mAccDiffMap of
+            Nothing -> return ()
+            Just accDiffMap -> do
+                flattenedAccounts <- liftIO $ DiffMap.flatten accDiffMap
+                LMDBAccountMap.insert flattenedAccounts
+                -- The difference map is set to @Nothing@ as any potential new accounts have been written to the
+                -- lmdb account map.
+                liftIO $ modifyIORef' accountDiffMap (const Nothing)
```

Comment:
Don't you need `atomicWriteIORef` instead? This variable can potentially be accessed by concurrent readers, can it not?

Reply:
Yes I need that.